### ISSUE TITLE
Lint on cloning &&T

### ISSUE THREAD

**Issue by Manishearth** [Manishearth/rust-clippy#620]
The following [code](http://is.gd/L8YHIa):

```rust
fn main() {
    let x = vec![1];
    let y = &&x;
    let z: &Vec<_> = y.clone();
    println!("{:p} {:p}",*y, z);
}
```

prints out the same pointer for `*y` and `z`. This is because `(&&T).clone()` will clone the `&T` pointer, not the `T`: because `&T` itself is `Clone` (being `Copy`), autoderef won't go any further. `&&T` is a common type that crops up when writing overly conservative code with `ref`s and borrows.

This probably could lead to some interesting bugs.

Might need to be careful about autoderef  in the implementation here, but I suspect it will be straightforward.

h/t @Gankro for pointing this out.

**Comment by Gankro** [comment]
Note that I'm at a loss for how this could make non-toy code do bad things. Either inference will push you to deref more, or you'll get a type error in 99% of cases. That said, if this can provide a better error message for those cases, great!

**Comment by Manishearth** [comment]
Yeah, I suspect the borrow checker will prevent most footguns here. However, `&&T` itself is something that commonly crops up so it's good to have a separate lint that points it out so that you can make sense of the borrow error.

> Either inference will push you to deref more

That can't happen. `(&&T).clone()` will _always_ return an `&T`, regardless of inference.
